src/main/java/org/spongepowered/common/entity/effect/VisibleVanishState.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.entity.effect;
+
+import org.spongepowered.api.effect.VanishState;
+
+public final class VisibleVanishState implements VanishState {
```

Comment:
Could you not remove this and simply set the flags on the previous impl? Is there truly an issue with adjusting the state off of this one (which produces a new state?)

Reply:
It just makes the Boolean logic a little unwieldy, so:
```java
    public VanishState affectMonsterSpawning(boolean affectsMonsterSpawning) {
        if (!this.vanished) {
            return this;
        }
        if (this.affectsSpawning == affectsMonsterSpawning) {
            return this;
        }
        return new SpongeVanishState(
            true,
            this.ignoresCollisions,
            this.untargetable,
            affectsMonsterSpawning,
            this.createsSounds,
            this.createsParticles
        );
    }
```
would be the odd thing of keeping the tristate (are we vanished) in the method, whereas the "Oh, we're a completely different class of vanish that's constant" makes it easier to reason about.